test/elk-kinesis-logger.js

```diff
@@ -0,0 +1,186 @@
+const AWS = require('aws-sdk-mock');
+const assert = require('assert');
+const MockDate = require('mockdate');
+const ELKKinesisLogger = require('../src/elk-kinesis-logger');
+
+describe('ELKKinesisLogger', () => {
+  const config = {
+    stage: 'TEST',
+    stack: 'elk-kinesis-logger',
+    app: 'elk-kinesis-logger-tests',
+    roleArn: 'test-role',
+    streamName: 'test-stream',
+    verbose: false
+  };
+
+  const logMsg = 'testing testing, is this thing on?';
+
+  const date = new Date(2017, 0, 1);
+
+  before(() => {
+    MockDate.set(date);
+
+    AWS.mock('STS', 'assumeRole', (params, callback) => {
+      callback(null, {
+        Credentials: {
+          AccessKeyId: 'secretID',
+          SecretAccessKey: 'secretKey',
+          SessionToken: 'secretToken'
+        }
+      });
+    });
+
+    AWS.mock('Kinesis', 'putRecord', (params, callback) => {
+      callback(null, {});
+    });
+  });
+
+  after(() => {
+    MockDate.reset();
+    AWS.restore();
+  });
+
+  it('should raise an error if not opened', () => {
+    const logger = new ELKKinesisLogger(config);
+
+    try {
+      logger.log('test');
+    } catch (actual) {
+      assert.equal(
+        actual,
+        'ELKKinesisLogger has not been opened. Be sure to call .open() first.'
+      );
+    }
+  });
+
+  it('should write a simple log to kinesis', done => {
+    const logger = new ELKKinesisLogger(config);
+
+    logger.open().then(() => {
+      logger.log(logMsg);
+
+      logger
+        .close()
+        .then(actual => {
+          const expected = [
+            {
+              stack: 'elk-kinesis-logger',
+              stage: 'TEST',
+              app: 'elk-kinesis-logger-tests',
+              timestamp: date,
+              level: 'INFO',
+              message: logMsg
+            }
+          ];
+
+          assert.deepEqual(actual, expected);
+          done();
+        })
+        .catch(err => done(new Error(err)));
+    });
+  });
+
+  it('should write a log with extra detail to kinesis', done => {
+    const logger = new ELKKinesisLogger(config);
+
+    logger.open().then(() => {
+      const extraDetail = {
+        foo: 'bar'
+      };
+
+      logger.log(logMsg, extraDetail);
+
+      logger
+        .close()
+        .then(actual => {
+          const expected = [
+            {
+              stack: 'elk-kinesis-logger',
+              stage: 'TEST',
+              app: 'elk-kinesis-logger-tests',
+              timestamp: date,
+              level: 'INFO',
+              message: logMsg,
+              foo: 'bar'
+            }
+          ];
+
+          assert.deepEqual(actual, expected);
```

Comment:
I've not been successful in doing this. I think its because I'm using https://github.com/dwyl/aws-sdk-mock and spies are hard... I'll continue on as its close to the end of the day, but would you be happy to pair on Monday otherwise?